src/Servers/Kestrel/Core/src/Internal/Http/HttpRequestStream.cs

```diff
@@ -227,5 +249,166 @@ private async Task CopyToAsyncInternal(Stream destination, CancellationToken can
                 }
             }
         }
+
+        private class AsyncPipeReader : IValueTaskSource<int>
+        {
+            private readonly PipeReader _pipeReader;
+            private readonly Action _onCompletedAction;
+            private readonly IAsyncEnumerator<int> _reader;
+
+            private Memory<byte> _buffer;
+            private CancellationToken _cancellationToken;
+            private ValueTask<bool> _readTask;
```

Comment:
Store the awaiter.

Reply:
Would make it harder to get a value for GetStatus then?

Reply:
Suppose I can grab the state upfront